src/dbnode/storage/index/compaction/compactor.go

```diff
@@ -0,0 +1,349 @@
+		return c.compactFromBuilderWithLock(builder)
+	}
+
+	// Need to combine segments first
+	batch := c.docsPool.Get()
+	defer func() {
+		c.docsPool.Put(batch)
+	}()
+
+	// flushBatch is declared to reuse the same code from the
+	// inner loop and the completion of the loop
+	flushBatch := func() error {
+		if len(batch) == 0 {
+			// Last flush might not have any docs enqueued
+			return nil
+		}
+
+		err := builder.InsertBatch(index.Batch{
+			Docs:                batch,
+			AllowPartialUpdates: true,
+		})
+		if err != nil && index.IsBatchPartialError(err) {
+			// If after filtering out duplicate ID errors
+			// there are no errors, then this was a successful
+			// insertion.
+			batchErr := err.(*index.BatchPartialError)
+			// NB(r): FilterDuplicateIDErrors returns nil
+			// if no errors remain after filtering duplicate ID
+			// errors, this case is covered in unit tests.
+			err = batchErr.FilterDuplicateIDErrors()
+		}
+		if err != nil {
+			return err
+		}
+
+		// Reset docs batch for reuse
+		batch = batch[:0]
+		return nil
+	}
+
+	for _, seg := range segs {
+		reader, err := seg.Reader()
+		if err != nil {
+			return nil, err
+		}
+
+		iter, err := reader.AllDocs()
+		if err != nil {
+			return nil, err
+		}
+
+		for iter.Next() {
+			batch = append(batch, iter.Current())
+			if len(batch) < c.docsMaxBatch {
+				continue
+			}
+			if err := flushBatch(); err != nil {
+				return nil, err
+			}
+		}
+
+		if err := iter.Err(); err != nil {
+			return nil, err
+		}
+		if err := iter.Close(); err != nil {
+			return nil, err
+		}
+		if err := reader.Close(); err != nil {
+			return nil, err
+		}
+	}
+
+	// Last flush in case some remaining docs that
+	// weren't written to the mutable segment
+	if err := flushBatch(); err != nil {
+		return nil, err
+	}
+
+	return c.compactFromBuilderWithLock(builder)
+}
+
+func (c *Compactor) compactFromBuilderWithLock(
+	builder segment.Builder,
+) (segment.Segment, error) {
+	defer func() {
+		// Release resources regardless of result,
+		// otherwise old compacted segments are held onto
+		// strongly
+		builder.Reset(0)
+	}()
+
+	// Since this builder is likely reused between compaction
+	// runs, we need to copy the docs slice
+	allDocs := builder.Docs()
+	if len(allDocs) == 0 {
+		return nil, errCompactorBuilderEmpty
+	}
+
+	allDocsCopy := make([]doc.Document, len(allDocs))
+	copy(allDocsCopy, allDocs)
+
+	err := c.writer.Reset(builder)
+	if err != nil {
+		return nil, err
+	}
+
+	success := false
+	closers := new(closers)
+	fstData := fst.SegmentData{
+		MajorVersion: c.writer.MajorVersion(),
+		MinorVersion: c.writer.MinorVersion(),
+		Metadata:     append([]byte(nil), c.writer.Metadata()...),
+		DocsReader:   docs.NewSliceReader(0, allDocsCopy),
+		Closer:       closers,
+	}
+
+	// Cleanup incase we run into issues
+	defer func() {
+		if !success {
+			closers.Close()
+		}
+	}()
+
+	c.buff.Reset()
+	if err := c.writer.WritePostingsOffsets(c.buff); err != nil {
```

Comment:
Yeah, would be pretty hard I think... wish there was like a mmap backed bytes.Buffer that grew by 2x and realloc'd each time.  Although that'd get expensive too since it's a syscall each time.

Reply:
Yeah...I'm like this close to writing a crappy `malloc` implementation to manage our mmaps so we can use them more aggressively lol